docs/source/notes/mkldnn.rst

```diff
@@ -0,0 +1,101 @@
+.. meta::
+   :description: A guide to torch.backends.mkldnn, a PyTorch backend to run MKLDNN operations
+   :keywords: optimize PyTorch, MKLDNN
+
+.. _mkldnn_backend:
+
+MKLDNN backend
+---------------------------------------------------
+
+MKLDNN is an open-source cross-platform performance library of basic building blocks
+for deep learning applications.
+
+.. code:: python
+
+  # The flag below controls whether enable MKLDNN backend in Pytorch.
+  torch.backends.mkldnn.enabled = True
+
+Users can disable MKLDNN backend by:
+
+.. code:: python
+
+  torch.backends.mkldnn.enabled = False
+
+.. _bf16_on_mkldnn:
+
+Bfloat16 (BF16) on MKLDNN backend
+---------------------------------------------------
+
+Starting in PyTorch 2.4, there is a set of APIs to control the internal computation precision
+for `float32` operators.
+
+.. code:: python
+
+  # The flag below controls the internal computation precision for mkldnn matmul. Default is float32.
+  torch.backends.mkldnn.matmul.fp32_precision = "default"
+
+  # The flag below controls the internal computation precision for mkldnn conv. Default is float32.
+  torch.backends.mkldnn.conv.fp32_precision = "default"
+
+  # The flag below controls the internal computation precision for mkldnn rnn. Default is float32.
+  torch.backends.mkldnn.rnn.fp32_precision = "default"
+
+Note that besides matmuls and convolutions themselves, functions and nn modules that internally uses
+matmuls or convolutions are also affected. These include `nn.Linear`, `nn.Conv*`, cdist, tensordot,
+affine grid and grid sample, adaptive log softmax, GRU and LSTM.
+
+To get an idea of the precision and speed, see the example code and benchmark data (on SPR) below:
+
+.. code:: python
+
+  torch.manual_seed(0)
+  a_full = torch.randn(10240, 10240, dtype=torch.double)
+  b_full = torch.randn(10240, 10240, dtype=torch.double)
+  ab_full = a_full @ b_full
+  mean = ab_full.abs().mean()  # 80.7451
+
+  a = a_full.float()
+  b = b_full.float()
+
+  # Do matmul at BF16 mode.
+  torch.backends.mkldnn.matmul.fp32_precision = 'bf16'
+  ab_bf16 = a @ b  # expected speedup with BF16 dot-product acceleration
+  error = (ab_bf16 - ab_full).abs().max()  # 1.3704
+  relative_error = error / mean  # 0.0170
+  print(error, relative_error)
+
+  # Do matmul FP32 mode.
+  torch.backends.mkldnn.matmul.fp32_precision = 'default'
+  ab_fp32 = a @ b
+  error = (ab_fp32 - ab_full).abs().max()  # 0.0003
+  relative_error = error / mean  # 0.00000317
+  print(error, relative_error)
+
+From the above example, we can see that with BF16, the speed is ~7x faster on SPR, and that
+relative error compared to double precision is approximately 2 orders of magnitude larger.
```

Comment:
I'm not sure what this section is about?
These notes are used to describe implementation details and hw specificities.
I don't think we want benchmarking code here. You can have this in a blogpost.

Reply:
We put it here because we wish to following CUDA TF32 https://github.com/pytorch/pytorch/blob/main/docs/source/notes/cuda.rst#tensorfloat-32-tf32-on-ampere-and-later-devices to demonstrate MKLDNN BF16 internal computations. The code snap just wish to show use BF16 as internal dtype will have some abs and relative error but also be faster. Users can follow this to do experiment on their machines if they are interested